lib/web/resources.go

```diff
@@ -364,4 +553,11 @@ type resourcesAPIGetter interface {
 	DeleteTrustedCluster(ctx context.Context, name string) error
 	// ListResoures returns a paginated list of resources.
 	ListResources(ctx context.Context, req proto.ListResourcesRequest) (*types.ListResourcesResponse, error)
+
+	GetApplicationServers(context.Context, string) ([]types.AppServer, error)
```

Comment:
WDYT about having a separate interface that combines resourcesAPIGetter and the methods below?

Or, better yet, smaller specializations for each resource-get method?

Reply:
I'm not sure it's worth the effort. This change will only go in v9 and v10, but not in v11+. But out of curiosity, what do you mean by smaller specializations for each getter? Like something like below:

```
type resourcesAPI interface {
	...
	ListResources(ctx context.Context, req proto.ListResourcesRequest) (*types.ListResourcesResponse, error)
	Ping(ctx context.Context) (proto.PingResponse, error)
}

type nodeAPI interface {
	GetNodes(ctx context.Context, namespace string, opts ...services.MarshalOption) ([]types.Server, error)
        // And all things related to node like create/delete?
}

type combinedApi interface {
    resourcesAPI
    nodeAPI
   // etc
}
```

I guess i'm confused what the benefit would be

Reply:
Yes, something like this:

```go
type resourcesAPIGetter interface {
	// as-is
}

type nodeGetter interface {
	resourcesAPIGetter
	GetNodes(ctx context.Context, namespace string, opts ...services.MarshalOption) ([]types.Server, error)
}

func handleClusterNodesGet(clt nodeGetter, r *http.Request, clusterName string, userRoles services.RoleSet) (*listResourcesGetResponse, error) {
	// as-is
}
```

The benefit here is that we don't (further) expand the `resourcesAPIGetter` interface. It's an attempt to keep the interfaces smaller, which is more in-line with Go best practices.

> This change will only go in v9 and v10

Is the plan to "forward-port" to v10?

Reply:
> Is the plan to "forward-port" to v10?

yes, is that the wrong approach? this is my first time fixing something in a later branch, and then working myself up to cover more compats. (since this feature was out 9.1, i have to cover 9.0) We didn't need this in v11...

Reply:
> The benefit here is that we don't (further) expand the resourcesAPIGetter interface. It's an attempt to keep the interfaces smaller, which is more in-line with Go best practices.

gotcha, i'll keep this in mind moving forward

Reply:
> yes, is that the wrong approach? this is my first time fixing something in a later branch, and then working myself up to cover more compats. (since this feature was out 9.1, i have to cover 9.0) We didn't need this in v11...

I suppose it could go either way. I think it's probably worth noting in the PR's description that it affects v10 too, just for completeness.

Reply:
got it, i did mention it in my PR description (under TODO) but i could've made it more clear